src/api/app/models/workflow/step/set_flags.rb

```diff
@@ -0,0 +1,82 @@
+class Workflow::Step::SetFlags < ::Workflow::Step
+  REQUIRED_KEYS = [:flags].freeze
+  REQUIRED_FLAG_KEYS = [:type, :status, :project].freeze
+  OPTIONAL_FLAG_KEYS = [:package, :repository, :architecture].freeze
+
+  validate :validate_flags
+
+  def call(_options = {})
+    return unless valid?
+
+    set_flags
+  end
+
+  def flags
+    step_instructions[:flags]
+  end
+
+  private
+
+  def set_flags
+    ActiveRecord::Base.transaction do
+      flags.each do |flag|
+        project = Project.find_by(name: target_project_name(project_name: flag[:project]))
+        raise ActiveRecord::RecordNotFound, "Couldn't find Project #{flag[:project]}" if project.nil?
+
+        package = project.packages.find_by(name: target_package_name(package_name: flag[:package])) if flag[:package].present?
+        main_object = package.presence || project
+        check_access(main_object)
+
+        architecture_id = Architecture.find_by_name(flag[:architecture]).id if flag[:architecture]
+        existing_flag = main_object.flags.exists?(flag: flag[:type], repo: flag[:repository], architecture_id: architecture_id)
+        next if existing_flag
+
+        main_object.add_flag(flag[:type], flag[:status], flag[:repository], flag[:architecture])
+        main_object.save!
+      end
+    end
+  end
+
+  def check_access(object)
+    raise Pundit::NotAuthorizedError unless Pundit.policy(token.user, object).update?
+  end
+
+  def validate_flags
+    return if flags.all? { |flag| (REQUIRED_FLAG_KEYS - flag.keys).empty? }
+
+    required_flag_keys_sentence ||= REQUIRED_FLAG_KEYS.map { |key| "'#{key}'" }.to_sentence
+    errors.add(:base, "set_flags step: All flags must have the #{required_flag_keys_sentence} keys")
+  end
+
+  # TODO: Totally duplicated from Workflow::Step. Remove the duplication by using a service instead for all steps depending on this method.
```

Comment:
We're tracking this refactoring in our [backlog](https://trello.com/c/WrOjCG1m/2490-remove-duplicated-methods-targetprojectname-and-targetpackagename).